Strips Offline code: TCDS-Scal migration

vischia opened the pull request:
This PR replaces SCAL access with TCDS access for the Strips offline code.

SCAL access is left as a fallback in case TCDS access fails.

The request is documented at https://its.cern.ch/jira/browse/CMSTRACK-166

# Tests:
I checked by adding printouts that the numerical values extracted from TCDSRecord are exactly the same as those extracted from L1ABC.
Some of these printouts are documented in https://github.com/vischia/cmssw/commit/73b5972c6af3d75cd235a9485e82b0d51fbc7d91

This is my first official PR to CMSSW and my first work on SiStrip; I tried to avoid bad practices (e.g. white lines addition/removal; I left only those which imho improve the logical readability of the code), I apologize in advance if there is something still not OK.

# Notes

- The bxoffset is defined as "eventBx - l1abcBx" in one module, and "l1abcBx - eventBx" in the other module. Since I have no idea of whether this is OK, I left it as it is, and I maintained this difference of conventions in the TCDS portion of the code.

Cheers,
Pietro

vischia committed:
Work in progress

vischia committed:
Work in progress

vischia committed:
Align for preparing PR

vischia committed:
Align for preparing PR

vischia committed:
Remove white line added by previous commit

vischia committed:
Remove white line added by previous commit

vischia committed:
EventWIthHistoryProducerFromL1ABC should be fine now

vischia committed:
EventWIthHistoryProducerFromL1ABC should be fine now

vischia committed:
TCDSRecord in APVCyclePhaseProducer

vischia committed:
TCDSRecord in APVCyclePhaseProducer

vischia committed:
Remove white lines

vischia committed:
Remove white lines

vischia committed:
Remove white lines

vischia committed:
Remove white lines

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14020

- This PR adds an extra 36KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14020/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14020/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

vischia committed:
Automatic patches

I don't understand why _tcdsRecordToken( and mayConsume must be on the same line, but the old _l1abccollectionToken( and mayConsume can remain on different lines.

vischia committed:
Automatic patches

I don't understand why _tcdsRecordToken( and mayConsume must be on the same line, but the old _l1abccollectionToken( and mayConsume can remain on different lines.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14023

- This PR adds an extra 36KB to repository

cmsbuild commented:
A new Pull Request was created by  @vischia (Pietro Vischia) for master.

It involves the following packages:

Configuration/StandardSequences
DPGAnalysis/SiStripTools


@silviodonato, @cmsbuild, @franzoni, @santocch, @fabiocos, @davidlange6 can you please review it and eventually sign? Thanks.
@echabert, @makortel, @felicepantaleo, @pieterdavid, @GiacomoSguazzoni, @rovere, @VinInn, @Martin-Grunewald, @lecriste, @jandrea, @ebrondol, @venturia, @mmusich, @dgulhan, @threus, @robervalwalsh this is something you requested to watch as well.
@davidlange6, @silviodonato, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mmusich commented on Configuration/StandardSequences/python/AlCaRecoStreams_cff.py:
```diff
@@ -175,6 +175,7 @@
 pathALCARECOSiStripCalMinBiasAAG = cms.Path(seqALCARECOSiStripCalMinBiasAAG*ALCARECOSiStripCalMinBiasAAGDQM)
 pathALCARECOSiStripCalSmallBiasScan = cms.Path(seqALCARECOSiStripCalSmallBiasScan)
 pathALCARECOSiStripCalZeroBias = cms.Path(seqALCARECOSiStripCalZeroBias*ALCARECOSiStripCalZeroBiasDQM)
+pathALCARECOSiStripCalZeroBias.replace(consecutiveHEs, unpackTcds*consecutiveHEs)
```
why is this done here, instead of here: https://github.com/cms-sw/cmssw/blob/master/Calibration/TkAlCaRecoProducers/python/ALCARECOSiStripCalZeroBias_cff.py#L56 ?

vischia replied:
Fixed (will push in a while)

vischia replied:
I thought it was better to limit the substitution to DPGAnalysis code, instead of changing it for all the Calibration package. 

Can you please confirm it is fine to do it in Calibration/...?

mmusich replied:
@vischia I am not sure to understand the argument about limiting the substitution to the DPGAnalysis code. Even here you are touching another package (Configuration/StandardSequences) and IMHO is better to change the subsystem packages (as Calibration) than the top-level configuration.

vischia replied:
@mmusich fair enough. I'll move the change to the relevant subsystem package.

mmusich commented on DPGAnalysis/SiStripTools/plugins/APVCyclePhaseProducerFromL1ABC.cc:
```diff
@@ -192,24 +198,43 @@ void APVCyclePhaseProducerFromL1ABC::produce(edm::Event& iEvent, const edm::Even
-          edm::LogWarning("L1AcceptBunchCrossingNoType") << "L1AcceptBunchCrossing with no type found: ";
-          for (L1AcceptBunchCrossingCollection::const_iterator debu = pIn->begin(); debu != pIn->end(); ++debu) {
-            edm::LogPrint("L1AcceptBunchCrossingNoType") << *debu;
+    try {
```
I think this goes against the code guidelines: http://cms-sw.github.io/cms_coding_rules.html (section 7, point 11)

vischia replied:
Fixed (will push in a while)

mmusich commented on DPGAnalysis/SiStripTools/plugins/EventWithHistoryProducerFromL1ABC.cc:
```diff
@@ -104,16 +107,24 @@ void EventWithHistoryProducerFromL1ABC::produce(edm::Event& iEvent, const edm::E
-          orbitoffset = (long long)l1abc->orbitNumber() - (long long)iEvent.orbitNumber();
-          bxoffset = l1abc->bunchCrossing() - iEvent.bunchCrossing();
+
+    try {
```
as above, I think we should try to avoid `try, catch(...)` statements.

vischia replied:
Fixed (will push in a while)

makortel commented on DPGAnalysis/SiStripTools/plugins/APVCyclePhaseProducerFromL1ABC.cc:
```diff
@@ -192,24 +198,43 @@ void APVCyclePhaseProducerFromL1ABC::produce(edm::Event& iEvent, const edm::Even
+        if (_hdorbit && *_hdorbit)
+          (*_hdorbit)->Fill(orbitoffset);
+      }
+    } catch (...) {
```
What throws the exception? Can it be modified to check the existence/validity explicitly upfront? (exceptions should be left to exceptional circumstances only, preferably left to framework to catch, and `catch(...)` should be avoided)

(same for all other try-catches)

mmusich replied:
same comment as https://github.com/cms-sw/cmssw/pull/29116#discussion_r388455950

makortel replied:
Right, we crossed.

vischia replied:
Fixed (will push in a while)

makortel commented on Configuration/StandardSequences/python/AlCaRecoStreams_cff.py:
```diff
@@ -175,6 +175,7 @@
 pathALCARECOSiStripCalMinBiasAAG = cms.Path(seqALCARECOSiStripCalMinBiasAAG*ALCARECOSiStripCalMinBiasAAGDQM)
 pathALCARECOSiStripCalSmallBiasScan = cms.Path(seqALCARECOSiStripCalSmallBiasScan)
 pathALCARECOSiStripCalZeroBias = cms.Path(seqALCARECOSiStripCalZeroBias*ALCARECOSiStripCalZeroBiasDQM)
+pathALCARECOSiStripCalZeroBias.replace(consecutiveHEs, unpackTcds*consecutiveHEs)
```
Why not modifying the sequence containing `consecutiveHEs` direclty?

mmusich replied:
same comment as https://github.com/cms-sw/cmssw/pull/29116#discussion_r388455489

vischia replied:
I thought it was better to replace that at runtime only for AlCaRecosStreams_cff.py.

Can you please confirm it is better to modify the original sequence directly?

vischia replied:
I have now seen comment https://github.com/cms-sw/cmssw/pull/29116#discussion_r388455489 . Fixed (will push in a while)

mmusich reviewed (commented):
@vischia, can you please update the title of the PR to make clear the changes are meant for Strips offline code? Also please remove the reference to JIRA (it's not clear which project you are referring to).
About the code changes I think the mandate of the ticket was to change also `DPGAnalysis/SiStripTools/plugins/APVCyclePhaseProducerFromL1TS.cc`. Do you foresee to work on that too?
I have some other comments below:

vischia committed:
Apply requested fixes

vischia committed:
Apply requested fixes

vischia commented:
I have also updated DPGAnalysis/SiStripTools/plugins/APVCyclePhaseProducerFromL1TS.cc, sorry I forgot to include it. I will push shortly when I'll have dealt with https://github.com/cms-sw/cmssw/pull/29116#discussion_r388455489

venturia commented:
.
> 
> SCAL access is left as a fallback in case TCDS access fails.
> 

A comment on the strategy described above.
Since both TCDS and SCAL were available during Run2 (but only SCAL was used and validated), my suggestion is to make all the plugins that are expected to choose between SCAL and TCDS configurable in order to force them to use SCAL even if TCDS is available. This will allow us to, for example,
- run two instances of these plugins, one configured to use SCAL and one configured to use TCDS, at the same time to compare event by event easily
- to use new releases on Run 2 data without being afraid that TCDS was not validated (I cannot exclude that there was a problem in TCDS data in Run 2)

Personally I would have approached the task differently by creating new plugins that access TCDS (only) and acted at the level of the configuration to choose which one has to be used.

vischia committed:
Also APVCyclePhaseProducerFromL1TS.cc

vischia committed:
Also APVCyclePhaseProducerFromL1TS.cc

cmsbuild commented:
The code-checks are being triggered in jenkins.

vischia commented:
@venturia , from what I understood in one of the RECO meetings it has been decided to use the fallback mechanism: https://its.cern.ch/jira/browse/CMSTRACK-166?focusedCommentId=2760041&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-2760041

venturia commented:
> @venturia , from what I understood in one of the RECO meetings it has been decided to use the fallback mechanism: https://its.cern.ch/jira/browse/CMSTRACK-166?focusedCommentId=2760041&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-2760041

Fine. But the possibility to force SCAL is needed. TCDS Run 2 data are NOT validated.

vischia commented:
> > @venturia , from what I understood in one of the RECO meetings it has been decided to use the fallback mechanism: https://its.cern.ch/jira/browse/CMSTRACK-166?focusedCommentId=2760041&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-2760041
> 
> Fine. But the possibility to force SCAL is needed. TCDS Run 2 data are NOT validated.

Fine; would a configuration parameter "forceSCAL" in the configuration file be OK?

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14054

- This PR adds an extra 40KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14054/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14054/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

venturia commented:
> > > @venturia , from what I understood in one of the RECO meetings it has been decided to use the fallback mechanism: https://its.cern.ch/jira/browse/CMSTRACK-166?focusedCommentId=2760041&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-2760041
> > 
> > 
> > Fine. But the possibility to force SCAL is needed. TCDS Run 2 data are NOT validated.
> 
> Fine; would a configuration parameter "forceSCAL" in the configuration file be OK?

In my opinion yes. But I let the software managers to comment about the details (default value, tracker vs untracked, ...)

vischia commented:
As a side comment; if things like

 _tcdsRecordToken(
                 mayConsume<TCDSRecord>(

are (and I agree on this) flagged as not OK, why when I fix the pre-existent 

 : _l1abccollectionToken(
          mayConsume<L1AcceptBunchCrossingCollection>(iConfig.getParameter<edm::In

to read


 : _l1abccollectionToken(mayConsume<L1AcceptBunchCrossingC....

does the change get rejected? What's the logic of it?

vischia commented:
> > > > @venturia , from what I understood in one of the RECO meetings it has been decided to use the fallback mechanism: https://its.cern.ch/jira/browse/CMSTRACK-166?focusedCommentId=2760041&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-2760041
> > > 
> > > 
> > > Fine. But the possibility to force SCAL is needed. TCDS Run 2 data are NOT validated.
> > 
> > 
> > Fine; would a configuration parameter "forceSCAL" in the configuration file be OK?
> 
> In my opinion yes. But I let the software managers to comment about the details (default value, tracker vs untracked, ...)

OK thanks! I'll wait on this then.

vischia committed:
Applied partially-nonsensical patch

vischia committed:
Applied partially-nonsensical patch

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14055

- This PR adds an extra 44KB to repository

cmsbuild commented:
Pull request #29116 was updated. @silviodonato, @cmsbuild, @franzoni, @santocch, @fabiocos, @davidlange6 can you please check and sign again.

makortel commented:
> does the change get rejected (see https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14054/code-format.patch ) ? What's the logic of it?

Can't tell without trying, it becomes what `clang-format` does based on https://github.com/cms-sw/cmssw/blob/master/.clang-format

makortel commented:
> But I let the software managers to comment about the details (default value, tracker vs untracked, ...)

In general the default values should correspond to Run 1, with customizations for later periods done with era modifiers. Parameters should be tracked if they can have any impact on results/physics.

makortel commented on DPGAnalysis/SiStripTools/plugins/APVCyclePhaseProducerFromL1ABC.cc:
```diff
@@ -186,30 +189,54 @@ void APVCyclePhaseProducerFromL1ABC::produce(edm::Event& iEvent, const edm::Even
+    Handle<TCDSRecord> tcds_pIn;
+    iEvent.getByToken(_tcdsRecordToken, tcds_pIn);
+    const auto& tcdsRecord = *tcds_pIn.product();
+    bool useTCDS(tcds_pIn.product());
```
Is the intention to use `useTCDS` to check whether a `TCDSRecord` exists or not?

vischia replied:
Yes. I could not find a better way; if you can suggest one it would be awesome!

About the clang-format, the point is that it seems it checks the rules only for the changes, and does not allow me to fix similarly wrong syntax that was already there in the module, which does not make sense to me.

makortel replied:
> Yes. I could not find a better way; if you can suggest one it would be awesome!

```
if(tcds_pIn.isValid()) {
   // TCDSRecord exists
} else {
  // TCDSRecord does not exist
}
```
The `tcds_pIn.product()` throws an exception if the product does not exist.

> About the clang-format, the point is that it seems it checks the rules only for the changes, and does not allow me to fix similarly wrong syntax that was already there in the module, which does not make sense to me.

Our format definition imposes maximum width for the line, and long lines get broken to multiple lines in some way. Therefore adding one character to a line can lead to reformatting of the line.

vischia replied:
Thank you very much!

mmusich commented on DPGAnalysis/SiStripTools/plugins/EventWithHistoryProducerFromL1ABC.cc:
```diff
@@ -104,16 +107,26 @@ void EventWithHistoryProducerFromL1ABC::produce(edm::Event& iEvent, const edm::E
+          if (l1abc->l1AcceptOffset() == 0) {
+            orbitoffset = (long long)l1abc->orbitNumber() - (long long)iEvent.orbitNumber();
+            bxoffset = l1abc->bunchCrossing() - iEvent.bunchCrossing();
+          }
```
@vischia I think we need also a new `EventWithHistory` constructor.
The one called below at line 134 in this file:
```C++
std::unique_ptr<EventWithHistory> pOut(new EventWithHistory(iEvent, *pIn, orbitoffset, bxoffset));
```
uses `Handle<L1AcceptBunchCrossingCollection> pIn`.
Now IIUC this object is put in the event by the SCAL unpacker here:
https://github.com/cms-sw/cmssw/blob/master/EventFilter/ScalersRawToDigi/src/ScalersRawToDigi.cc#L55.
This means that if SCAL is not in the DAQ the object used here will be empty.

vischia replied:
Oh, I see. Thanks for spotting @mmusich , I'll try to fix this on Monday.

vischia replied:
Hi @mmusich , I am adding the constructor, but I can't find the equivalent in https://github.com/cms-sw/cmssw/blob/master/DataFormats/TCDS/interface/TCDSRecord.h of the l1acceptOffset from https://github.com/cms-sw/cmssw/blob/master/DataFormats/Scalers/interface/L1AcceptBunchCrossing.h .

While in the other modules I solved by just removing the check ```if l1abc->l1AcceptOffset() == 0)```, https://github.com/cms-sw/cmssw/blob/master/DPGAnalysis/SiStripTools/src/EventWithHistory.cc#L37 requires to compute a quantity and use it later, so it's not clear to me how to replace it with something from TCDSRecord

mmusich replied:
Finding out how to deal with it is part of the ticket mandate.

mmusich replied:
I haven't checked the details, but maybe this https://github.com/cms-sw/cmssw/blob/master/DataFormats/TCDS/interface/TCDSRecord.h#L109 can help?

vischia committed:
Moved sequence manipulation to submodules

vischia committed:
Moved sequence manipulation to submodules

vischia commented:
So, I added a commit which solves this: https://github.com/cms-sw/cmssw/pull/29116#discussion_r388455489 by moving the definition where suggested (and in another couple places, after testing with runTheMatrix).

On Monday I'll look into the two other points that have been raised (I hope I did not miss any other point):

- add a forceSCAL flag
- add a new EventWithHistory constructor which takes in a tcds record instead of the L1AcceptBunchCrossingCollection thing.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14060

- This PR adds an extra 40KB to repository

cmsbuild commented:
Pull request #29116 was updated. @silviodonato, @christopheralanwest, @tocheng, @cmsbuild, @franzoni, @tlampen, @pohsun, @santocch, @fabiocos, @davidlange6 can you please check and sign again.

vischia committed:
Added new constructor in EventWithHistory, accepting TCDSRecord as input

vischia committed:
Added new constructor in EventWithHistory, accepting TCDSRecord as input

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14089

- This PR adds an extra 24KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14089/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14089/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

vischia commented:
I fixed what was requested:

- use isValid() for checking whether the tcdsRecord is valid
- add a new EventWithHistory constructor which takes in a tcds record instead of the L1AcceptBunchCrossingCollection thing.

I think the only missing item is

- add a forceSCAL flag, but the latest I have read on this is that we have to wait for some expert to mention which type (tracked? untracked? etc) parameter do we need https://github.com/cms-sw/cmssw/pull/29116#issuecomment-595771960

vischia committed:
Code-checks

vischia committed:
Code-checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14090

- This PR adds an extra 92KB to repository

cmsbuild commented:
Pull request #29116 was updated. @silviodonato, @christopheralanwest, @tocheng, @cmsbuild, @franzoni, @tlampen, @pohsun, @santocch, @fabiocos, @davidlange6 can you please check and sign again.

makortel commented:
> * add a forceSCAL flag, but the latest I have read on this is that we have to wait for some expert to mention which type (tracked? untracked? etc) parameter do we need [#29116 (comment)](https://github.com/cms-sw/cmssw/pull/29116#issuecomment-595771960)

Parameters should be tracked if they can have any impact on results/physics https://github.com/cms-sw/cmssw/pull/29116#issuecomment-595819265. So it should be tracked.

makortel commented on DPGAnalysis/SiStripTools/plugins/APVCyclePhaseProducerFromL1ABC.cc:
```diff
@@ -186,30 +189,59 @@ void APVCyclePhaseProducerFromL1ABC::produce(edm::Event& iEvent, const edm::Even
+    Handle<TCDSRecord> tcds_pIn;
+    iEvent.getByToken(_tcdsRecordToken, tcds_pIn);
+    const auto& tcdsRecord = *tcds_pIn.product();
+    bool useTCDS(tcds_pIn.isValid());
```
The `tcds_pIn.product()` still throws an exception when the `TCDSRecord` does not exist because it is called also when `useTCDS == false`.

vischia replied:
Fixed.

vischia committed:
Added parameter to force SCAL

vischia committed:
Added parameter to force SCAL

cmsbuild commented:
The code-checks are being triggered in jenkins.

mmusich commented on DPGAnalysis/SiStripTools/python/apvcyclephaseproducerfroml1ts2010_cfi.py:
```diff
@@ -9,5 +9,7 @@
                            defaultPhases = cms.vint32(30,30,30,30),
                            l1TSCollection = cms.InputTag("scalersRawToDigi"),
+                           tcdsRecordLabel= cms.InputTag("unpackTcds","tcdsRecord"),
+                           forceSCAL = cms.bool(False),
```
at @vischia quoting Matti here:

> In general the default values should correspond to Run 1, with customizations for later periods done with era modifiers. Parameters should be tracked if they can have any impact on results/physics.

Isn't this breaking the rule?

vischia replied:
From what I understand in Run1 TCDS was not available, so SCAL will be used anyways, regardless of the value of forceSCAL.

The purpose of forceSCAL is to force usage of SCAL when TCDS is available, not of pointing out whether TCDS was available or not.

mmusich replied:
I think what Andrea had in mind is something different. We must ensure that Run1 and Run2 do use SCAL no matter what.

vischia replied:
OK, so I put it to True by default, and then at RunIII one can activate it, would that be OK?

mmusich replied:
sounds good to me. Thanks!

vischia replied:
Or I can put it to something that gives True for Run1 and Run2, and false for Run3, if there is anything with this behaviour. How do I access Run information at the level of the config files?

vischia replied:
> sounds good to me. Thanks!

Oh OK, then I'll do that, thanks!

mmusich replied:
I think the default should be to true, and using the Run3 era modifier it should be allowed to take TCDS when available.

mommsen replied:
Note that SCAL will no longer exist in run 3. Thus, you have to use TCDS in the future.

mmusich replied:
@mommsen that's why in the run3 era that parameter is set to False:
https://github.com/cms-sw/cmssw/blob/9cb7235ee1351aed1f9a8f2a15c9b6d435e495f0/DPGAnalysis/SiStripTools/python/apvcyclephaseproducerfroml1ts_cfi.py#L16-L17

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14091

- This PR adds an extra 100KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14091/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14091/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

makortel commented on DPGAnalysis/SiStripTools/plugins/APVCyclePhaseProducerFromL1ABC.cc:
```diff
@@ -186,30 +190,59 @@ void APVCyclePhaseProducerFromL1ABC::produce(edm::Event& iEvent, const edm::Even
+    Handle<TCDSRecord> tcds_pIn;
+    iEvent.getByToken(_tcdsRecordToken, tcds_pIn);
+    bool useTCDS(tcds_pIn.isValid() && !_forceSCAL);
+    const auto& tcdsRecord = useTCDS ? *tcds_pIn.product() : NULL;
```
Reference to `nullptr` is undefined behavior.

vischia replied:
When it is null it is not accessed (null and access are both regulated by useTCDS), I don't understand what's the problem

vischia replied:
Ah, you do mean auto& of NULL?

makortel replied:
It is undefined behavior according to the language, so the compiler is free to do literally anything (even if not today, it may tomorrow). If you can't follow the pattern
```cpp
if (tcds_pIn.isValid()) {
   const auto& tcdsRecord = *tcds_pIn;
}
```
please use a pointer instead of a reference
```cpp
const auto* tcdsRecrod = useTCDS ? tcds_pIn.product() : nullptr;
```

vischia replied:
Fun fact: it compiled without any issue

makortel replied:
Lot's of undefined behavior compile fine.

vischia committed:
Changing default for forceSCAL (run1 and run2 must have it activated)

vischia committed:
Changing default for forceSCAL (run1 and run2 must have it activated)

cmsbuild commented:
The code-checks are being triggered in jenkins.

vischia committed:
Access to tcdsRecord by pointer

vischia committed:
Access to tcdsRecord by pointer

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14093

- This PR adds an extra 32KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14093/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14093/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

mmusich commented on Configuration/StandardSequences/python/AlCaRecoStreams_cff.py:
```diff
@@ -175,6 +175,7 @@
 pathALCARECOSiStripCalMinBiasAAG = cms.Path(seqALCARECOSiStripCalMinBiasAAG*ALCARECOSiStripCalMinBiasAAGDQM)
 pathALCARECOSiStripCalSmallBiasScan = cms.Path(seqALCARECOSiStripCalSmallBiasScan)
 pathALCARECOSiStripCalZeroBias = cms.Path(seqALCARECOSiStripCalZeroBias*ALCARECOSiStripCalZeroBiasDQM)
+#pathALCARECOSiStripCalZeroBias.replace(consecutiveHEs, unpackTcds*consecutiveHEs)
```
@vischia if this no more needed, please remove the commented code.

vischia replied:
Fixed

vischia committed:
scram code checks thing

vischia committed:
scram code checks thing

mmusich commented on DPGAnalysis/SiStripTools/python/apvcyclephaseproducerfroml1ts_cfi.py:
```diff
@@ -9,5 +9,7 @@
                            defaultPhases = cms.vint32(30,30,30,30),
                            l1TSCollection = cms.InputTag("scalersRawToDigi"),
+                           tcdsRecordLabel= cms.InputTag("unpackTcds","tcdsRecord"),
+                           forceSCAL = cms.bool(True),
```
@vischia I think here we should add:
```
from Configuration.Eras.Modifier_run3_common_cff import run3_common
run3_common.toModify(APVPhases, forceSCAL = False)
```

vischia replied:
OK thanks, I'll add it straight away!

mmusich commented on DPGAnalysis/SiStripTools/python/eventwithhistoryproducerfroml1abc_cfi.py:
```diff
@@ -1,5 +1,10 @@
-                                l1ABCCollection=cms.InputTag("scalersRawToDigi")
+                                l1ABCCollection=cms.InputTag("scalersRawToDigi"),
+                                tcdsRecordLabel= cms.InputTag("unpackTcds","tcdsRecord"),
+                                forceSCAL = cms.bool(True)
```
@vischia I think here we should add:
```
from Configuration.Eras.Modifier_run3_common_cff import run3_common
run3_common.toModify(consecutiveHEs, forceSCAL = False)
```

vischia committed:
Removed commented line

vischia committed:
Removed commented line

cmsbuild commented:
The code-checks are being triggered in jenkins.

vischia committed:
Added automatic configuration for run3

vischia committed:
Added automatic configuration for run3

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14096

- This PR adds an extra 100KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14096/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14096/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

vischia commented:
I don't think the suggested patch makes sense, I won't apply it.

Is there a way of fixing or overriding that buggy check?

```
--- a/DPGAnalysis/SiStripTools/plugins/APVCyclePhaseProducerFromL1ABC.cc
+++ b/DPGAnalysis/SiStripTools/plugins/APVCyclePhaseProducerFromL1ABC.cc
@@ -203,7 +203,10 @@ void APVCyclePhaseProducerFromL1ABC::produce(edm::Event& iEvent, const edm::Even
       if (tcdsRecord->getL1aHistoryEntry(0).getIndex() == 0) {
         if (tcdsRecord->getEventType() != 0) {
           orbitoffset = (long long)tcdsRecord->getOrbitNr() - (long long)iEvent.orbitNumber();
-          bxoffset = iEvent.bunchCrossing()-tcdsRecord->getBXID();  // In EventWithHistoryProducerFromL1ABC, it's -1\times this, following the corresponding L1ABC line below. I keep this difference here, because I am not sure if it should be kept or not
+          bxoffset =
+              iEvent.bunchCrossing() -
+              tcdsRecord
+                  ->getBXID();  // In EventWithHistoryProducerFromL1ABC, it's -1\times this, following the corresponding L1ABC line below. I keep this difference here, because I am not sure if it should be kept or not
           // If I understand correctly, tcdsRecord has no l1AcceptOffset thing, so the control done for l1abc can be safely skipped
           if (_wantHistos) {
             if (_hbx && *_hbx)
```

makortel commented:
Move the comment to the preceding line.

mmusich commented on Calibration/TkAlCaRecoProducers/python/ALCARECOSiStripCalZeroBiasHI_cff.py:
```diff
@@ -59,4 +59,5 @@
 
 # Sequence #
-seqALCARECOSiStripCalZeroBias = cms.Sequence(ALCARECOSiStripCalZeroBiasHLT*HLTPixelActivityFilterForSiStripCalZeroBias*DCSStatusForSiStripCalZeroBias*calZeroBiasClusters*APVPhases*consecutiveHEs)
+from Configuration.StandardSequences.SimL1EmulatorRepack_Full_cff import unpackTcds
```
@vischia as far as I can tell, this line is not needed, right?

vischia replied:
@mmusich yes it's needed; the following line adds ```unpackTcds``` to the ```seqALCARECOSiStripCalZeroBias``` sequence

mmusich replied:
@vischia
I just tried without it and wf 1001.0 runs without issues. I think that is already imported by this, https://github.com/cms-sw/cmssw/blob/45b8639cbfd3e43caf9a9d6ef5f75395bea19a26/Calibration/TkAlCaRecoProducers/python/ALCARECOSiStripCalZeroBiasHI_cff.py#L51

vischia replied:
@mmusich  uh let me also check

vischia replied:
Yup, you are right. I fixed it now.

mmusich commented on Calibration/TkAlCaRecoProducers/python/ALCARECOSiStripCalZeroBias_cff.py:
```diff
@@ -53,7 +53,9 @@
 
 # Sequence #
-seqALCARECOSiStripCalZeroBias = cms.Sequence(ALCARECOSiStripCalZeroBiasHLT*DCSStatusForSiStripCalZeroBias*calZeroBiasClusters*APVPhases*consecutiveHEs)
+from Configuration.StandardSequences.SimL1EmulatorRepack_Full_cff import unpackTcds
```
@vischia as far as I can tell, this line is not needed, right?

vischia committed:
Fixed formatting

vischia committed:
Fixed formatting

vischia commented:
> Move the comment to the preceding line.

Oh is that the issue which breaks the check? OK, done! Thanks!

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14113

- This PR adds an extra 96KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14113/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14113/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

vischia committed:
Remove unnecessary import (already imported by other import statement)

vischia committed:
Remove unnecessary import (already imported by other import statement)

vischia committed:
Applied code sanity check patch

vischia committed:
Applied code sanity check patch

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14114

- This PR adds an extra 96KB to repository

cmsbuild commented:
Pull request #29116 was updated. @christopheralanwest, @tocheng, @cmsbuild, @franzoni, @tlampen, @pohsun, @santocch can you please check and sign again.

mmusich commented:
I think this should be ready for testing.
Would it be possible to include in the testing parameters for this PR wf 138.1 (which AFAIK tests 2020 MWGR data, without SCAL)?

vischia commented:
> I think this should be ready for testing.
> Would it be possible to include in the testing parameters for this PR wf 138.1 (which AFAIK tests 2020 MWGR data, without SCAL)?

@mmusich sorry for the naïve question (first PR for me): is this something I should do by writing something in the PR message, or something done by admins?

mmusich commented:
> is this something I should do by writing some tag in the PR message, or something done by admins?

It was a request for the CMSSW L2s in charge of the review:  @christopheralanwest, @tocheng,  @tlampen, @pohsun, @santocch

christopheralanwest commented:
test parameters
workflow = 138.1

christopheralanwest commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **MATRIX_EXTRAS** = 138.1

cmsbuild commented:
-1

Tested at: 1e10cc1d40144dd22f6bcbbaddba3a06f650bf5c

CMSSW: CMSSW_11_1_X_2020-03-10-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-524d9c/5102/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
1330.0 step3
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM+NANOUP15/step3_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM+NANOUP15.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

mmusich commented on Calibration/TkAlCaRecoProducers/python/ALCARECOSiStripCalZeroBiasHI_cff.py:
```diff
@@ -59,4 +59,5 @@
 
 # Sequence #
-seqALCARECOSiStripCalZeroBias = cms.Sequence(ALCARECOSiStripCalZeroBiasHLT*HLTPixelActivityFilterForSiStripCalZeroBias*DCSStatusForSiStripCalZeroBias*calZeroBiasClusters*APVPhases*consecutiveHEs)
+######from Configuration.StandardSequences.SimL1EmulatorRepack_Full_cff import unpackTcds
```
@vischia please remove commented line.

vischia replied:
> @vischia please remove commented line.

Uh, I must have taken it out only from one of the cffs, forgetting the others. Taking it out.

mmusich commented on Calibration/TkAlCaRecoProducers/python/ALCARECOSiStripCalZeroBias_cff.py:
```diff
@@ -53,7 +53,9 @@
 
 # Sequence #
-seqALCARECOSiStripCalZeroBias = cms.Sequence(ALCARECOSiStripCalZeroBiasHLT*DCSStatusForSiStripCalZeroBias*calZeroBiasClusters*APVPhases*consecutiveHEs)
+from Configuration.StandardSequences.SimL1EmulatorRepack_Full_cff import unpackTcds
```
@vischia why is this line still here?

vischia replied:
Uh, I must have taken it out only from one of the cffs, forgetting the others. Taking it out.

mmusich commented on DPGAnalysis/SiStripTools/python/eventwithhistoryproducerfroml1abc_cfi.py:
```diff
@@ -1,5 +1,12 @@
 import FWCore.ParameterSet.Config as cms
 
+import EventFilter.Utilities.tcdsRawToDigi_cfi
+from Configuration.StandardSequences.SimL1EmulatorRepack_Full_cff import unpackTcds
```
@vischia I think there is no need to unpack again TCDS, as it is done already centrally for everybody. We just need to use the correct product labels:
```
tcdsRecordLabel= cms.InputTag("tcdsDigis","tcdsRecord"),
```

vischia replied:
OK

mmusich reviewed (commented):
@vischia 
I think there are still problem with the unpacking of TCDS, which we don't need to redo again.
Can you please cherry-pick this commit: https://github.com/vischia/cmssw/commit/f373a6251029207d48f2883234088d67ea55cf67 ?

arossi83 commented:
@vischia I tested the SiStrip DQM online client:
I get the following message for every event

`%MSG-w L1AcceptBunchCrossingNoCollection:   EventWithHistoryProducerFromL1ABC:consecutiveHEs 10-Mar-2020 14:31:06 CET  Run: 335507 Event: 310532
 L1AcceptBunchCrossing with offset=0 not found  likely TCDSRecord is empty`

if it could help my guess is that this condition is never true https://github.com/cms-sw/cmssw/pull/29116/files?file-filters%5B%5D=.cc#diff-649c2d7f0bae523b9f652e9f8fb1f606R151

vischia commented:
> @vischia I tested the SiStrip DQM online client:
> I get the following message for every event
> 
> `%MSG-w L1AcceptBunchCrossingNoCollection: EventWithHistoryProducerFromL1ABC:consecutiveHEs 10-Mar-2020 14:31:06 CET Run: 335507 Event: 310532 L1AcceptBunchCrossing with offset=0 not found likely TCDSRecord is empty`
> 
> if it could help my guess is that this condition is never true https://github.com/cms-sw/cmssw/pull/29116/files?file-filters%5B%5D=.cc#diff-649c2d7f0bae523b9f652e9f8fb1f606R151

Since that condition is there in analogy to the other constructor (the one which takes L1abc as input), my guess is that l1AcceptOffset is loaded incorrectly; I load it from tcdsRecord.getL1aHistoryEntry(0).getIndex(), but maybe taking always the first entry is wrong. I'll try to debug it.

vischia commented:
> @vischia
> I think there are still problem with the unpacking of TCDS, which we don't need to redo again.
> Can you please cherry-pick this commit: [vischia@f373a62](https://github.com/vischia/cmssw/commit/f373a6251029207d48f2883234088d67ea55cf67) ?

Do you mean to revert it?

mmusich commented:
> Do you mean to revert it?

no, to apply it (which is reverting part of your changes + changing the source of the TCDS digis)

vischia committed:
Removed reunpacking

vischia committed:
Removed reunpacking

cmsbuild commented:
The code-checks are being triggered in jenkins.

mmusich commented on DPGAnalysis/SiStripTools/python/eventwithhistoryproducerfroml1abc_cfi.py:
```diff
@@ -1,5 +1,11 @@
 import FWCore.ParameterSet.Config as cms
 
+import EventFilter.Utilities.tcdsRawToDigi_cfi
```
since the source of TCDS digis is change this line is superfluous.

vischia replied:
Fixed (testing)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14133

- This PR adds an extra 100KB to repository

vischia commented:
> > @vischia I tested the SiStrip DQM online client:
> > I get the following message for every event
> > `%MSG-w L1AcceptBunchCrossingNoCollection: EventWithHistoryProducerFromL1ABC:consecutiveHEs 10-Mar-2020 14:31:06 CET Run: 335507 Event: 310532 L1AcceptBunchCrossing with offset=0 not found likely TCDSRecord is empty`
> > if it could help my guess is that this condition is never true https://github.com/cms-sw/cmssw/pull/29116/files?file-filters%5B%5D=.cc#diff-649c2d7f0bae523b9f652e9f8fb1f606R151
> 
> Since that condition is there in analogy to the other constructor (the one which takes L1abc as input), my guess is that ´´´l1AcceptOffset´´´ is loaded incorrectly; I load it from ´´´tcdsRecord.getL1aHistoryEntry(0).getIndex()´´´, but maybe taking always the first entry is wrong. I'll try to debug it.

From github search, git search, google, and the CMS lxr, it seems TCDSRecord::getL1aHistoryEntry is not used anywhere. 
The only thing that is used is the getFulll1aHistory, here https://github.com/cms-sw/cmssw/pull/25973/files#diff-442431741704aece7df5dde6a57f391cR143-R148 . I'll try to access that one and see how many entries does it have, and try to pick a meaningful one.

cmsbuild commented:
Pull request #29116 was updated. @cmsbuild, @santocch can you please check and sign again.

vischia committed:
Removed line

vischia committed:
Removed line

cmsbuild commented:
The code-checks are being triggered in jenkins.

mmusich commented on DPGAnalysis/SiStripTools/python/apvcyclephaseproducerfroml1ts_cfi.py:
```diff
@@ -9,5 +9,9 @@
                                                                ),
                            defaultPhases = cms.vint32(30,30,30,30),
                            l1TSCollection = cms.InputTag("scalersRawToDigi"),
+                           tcdsRecordLabel= cms.InputTag("unpackTcds","tcdsRecord"),
```
forgot to mention, but here too the source of TCDS digis needs to be modified. Sorry for not pointing it out earlier.

vischia replied:
No problem, sorry I should have remembered it was set also here

mmusich commented on DPGAnalysis/SiStripTools/python/apvcyclephaseproducerfroml1ts2010_cfi.py:
```diff
@@ -9,5 +9,9 @@
                                                                ),
                            defaultPhases = cms.vint32(30,30,30,30),
                            l1TSCollection = cms.InputTag("scalersRawToDigi"),
+                           tcdsRecordLabel= cms.InputTag("unpackTcds","tcdsRecord"),
```
and here too.

mmusich commented on DPGAnalysis/SiStripTools/python/apvcyclephaseproducerfroml1ts2011_cfi.py:
```diff
@@ -10,5 +10,9 @@
                            defaultPhases = cms.vint32(60,60,60,60),
                            magicOffset = cms.untracked.int32(258),
                            l1TSCollection = cms.InputTag("scalersRawToDigi"),
+                           tcdsRecordLabel= cms.InputTag("unpackTcds","tcdsRecord"),
```
and here.

mmusich commented on DPGAnalysis/SiStripTools/python/apvcyclephaseproducerfroml1ts2013_cfi.py:
```diff
@@ -10,5 +10,9 @@
                            defaultPhases = cms.vint32(48,48,48,48),
                            magicOffset = cms.untracked.int32(11),
                            l1TSCollection = cms.InputTag("scalersRawToDigi"),
+                           tcdsRecordLabel= cms.InputTag("unpackTcds","tcdsRecord"),
```
and here.

mmusich commented on DPGAnalysis/SiStripTools/python/apvcyclephaseproducerfroml1tsDB_cfi.py:
```diff
@@ -8,5 +8,9 @@
                            defaultPhases = cms.vint32(30,30,30,30),
                            l1TSCollection = cms.InputTag("scalersRawToDigi"),
-                           )
+                           tcdsRecordLabel= cms.InputTag("unpackTcds","tcdsRecord"),
```
and here.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14134

- This PR adds an extra 100KB to repository

cmsbuild commented:
Pull request #29116 was updated. @cmsbuild, @santocch can you please check and sign again.

vischia commented:
So, I grepped the tcdsRecordLabel and changed it in a few more places, committing shortly.

vischia committed:
Removed repeated unpacking, left only the main unpacking in SimL1EmulatorRepack_*

vischia committed:
Removed repeated unpacking, left only the main unpacking in SimL1EmulatorRepack_*

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14135

- This PR adds an extra 104KB to repository

cmsbuild commented:
Pull request #29116 was updated. @cmsbuild, @santocch can you please check and sign again.

mmusich commented:
@vischia thanks for following up to the review.
It would be nice to confirm the issue spotted by @arossi83 (https://github.com/cms-sw/cmssw/pull/29116#issuecomment-597173395) by running a new set of tests with the current status of the branch.

vischia commented:
> @vischia thanks for following up to the review.
> It would be nice to confirm the issue spotted by @arossi83 ([#29116 (comment)](https://github.com/cms-sw/cmssw/pull/29116#issuecomment-597173395)) by running a new set of tests with the current status of the branch.

@mmusich that's what I did in the last half an hour :)

As I mentioned in the answers, it's probably an issue of the `TCDSRecord` access to the l1a history to get the `l1AcceptOffset`.
I was guessing the first entry would be the one needed, but it turns out it makes the condition always fail.
I have now tried to access the whole l1accept history using `getFulll1aHistory`, in a similar way to  https://github.com/cms-sw/cmssw/pull/25973/files#diff-442431741704aece7df5dde6a57f391cR143-R148 but the workflow I am using (136.837) forces SCAL to false. Instead of changing its value by hand, I wanted to try the 138.1 mentioned in the tests, but I don't think it runs (it prints the messages below in less than 1sec and exits): is there any particular reason it can't run locally?

```
[pvischia@ingrid-ui1 src]$ runTheMatrix.py --nThreads=16 -l 138.1
processing relval_standard
processing relval_highstats
processing relval_pileup
processing relval_generator
processing relval_extendedgen
processing relval_production
processing relval_ged
ignoring relval_upgrade from default matrix
processing relval_2017
processing relval_2023
ignoring relval_identity from default matrix
processing relval_machine
processing relval_unsch
processing relval_premix
Running in 4 thread(s)
 tests passed,  failed
```

vischia commented:
OK, there seem to be always 16 records.

Givent that the first one experimentally did not work, resulting in an offset which made the conditional indicated by @arossi83 fail.
I guess it's one of the other ones, but I am searching for documentation about what they mean; maybe they represent previous offsets in order, and I have to take always "the previous one", which might be the index 1? Or is it too wild a guess?

```
TCDS RECORD L1A HISTORY: scanning content 
TCDS RECORD L1A HISTORY: 0, content 
TCDS RECORD L1A HISTORY: 1, content 
TCDS RECORD L1A HISTORY: 2, content 
TCDS RECORD L1A HISTORY: 3, content 
TCDS RECORD L1A HISTORY: 4, content 
TCDS RECORD L1A HISTORY: 5, content 
TCDS RECORD L1A HISTORY: 6, content 
TCDS RECORD L1A HISTORY: 7, content 
TCDS RECORD L1A HISTORY: 8, content 
TCDS RECORD L1A HISTORY: 9, content 
TCDS RECORD L1A HISTORY: 10, content 
TCDS RECORD L1A HISTORY: 11, content 
TCDS RECORD L1A HISTORY: 12, content 
TCDS RECORD L1A HISTORY: 13, content 
TCDS RECORD L1A HISTORY: 14, content 
TCDS RECORD L1A HISTORY: 15, content 
```

venturia commented:
While you look for the documentation my suggestion is to dump the L1ABC and TCDS content of a Run2 event when both were available

vischia commented:
> While you look for the documentation

In the twiki I could not find anything.

> my suggestion is to dump the L1ABC and TCDS content of a Run2 event when both were available

Yes I thought about it but I couldn't find any information on how can I know which events have both L1ABC and TCDS and on how to find the list of events of all runTheMatrix workflows to compare it with the list of events/runs with both L1ABC and TCDS.

I'll take a look and make some more searches next week or so.

mmusich commented:
@vischia 

> wanted to try the 138.1 mentioned in the tests (triggering the run3 condition, if I understand correctly), but I don't think it runs (it prints the messages below in less than 1sec and exits): is there any particular reason it can't run locally?

It runs just fine for me.

mmusich commented:
@vischia which is your release base? That workflow is defined only in CMSSW >= 11_0_X (introduced for the first time in https://github.com/cms-sw/cmssw/pull/27341)

mmusich commented:
@vischia 

> Yes I thought about it but I couldn't find any information on how can I know which events have both L1ABC and TCDS.

Any of the 2018 events has both SCAL and TDCS information (it is actually written in the JIRA ticket assignement). You can use e.g. 136.868 (RunZeroBias2018B+HLTDR2_2018+RECODR2_2018reHLT_ZBOffline+HARVEST2018ZB )

tsusa commented:
@vischia
Below is the dump of L1ABC and TCDS L1history for Run 331483, Event 3668020
L1ABC
AcceptOffset = 0, OrbitNr = 74005787, BXID = 3076, EventType = 1
AcceptOffset = -1, OrbitNr = 74005772, BXID = 2798, EventType = 1
AcceptOffset = -2, OrbitNr = 74005770, BXID = 2786, EventType = 1
AcceptOffset = -3, OrbitNr = 74005748, BXID = 743, EventType = 1
AcceptOffset = -4, OrbitNr = 74005742, BXID = 2380, EventType = 1
AcceptOffset = -5, OrbitNr = 74005740, BXID = 3490, EventType = 1
AcceptOffset = -6, OrbitNr = 74005719, BXID = 2537, EventType = 1
AcceptOffset = -7, OrbitNr = 74005699, BXID = 1941, EventType = 1
TCDS L1history 
index = -1, OrbitNr = 74005772, BXID = 2798, EventType = 3
index = -2, OrbitNr = 74005770, BXID = 2786, EventType = 3
index = -3, OrbitNr = 74005748, BXID = 743, EventType = 3
index = -4, OrbitNr = 74005742, BXID = 2380, EventType = 3
index = -5, OrbitNr = 74005740, BXID = 3490, EventType = 2
index = -6, OrbitNr = 74005719, BXID = 2537, EventType = 3
index = -7, OrbitNr = 74005699, BXID = 1941, EventType = 3
index = -8, OrbitNr = 74005694, BXID = 3039, EventType = 3
index = -9, OrbitNr = 74005686, BXID = 3506, EventType = 3
index = -10, OrbitNr = 74005656, BXID = 3325, EventType = 3
index = -11, OrbitNr = 74005628, BXID = 3490, EventType = 2
index = -12, OrbitNr = 74005558, BXID = 3021, EventType = 3
index = -13, OrbitNr = 74005517, BXID = 3263, EventType = 3
index = -14, OrbitNr = 74005516, BXID = 3490, EventType = 2
index = -15, OrbitNr = 74005459, BXID = 3293, EventType = 3
index = -16, OrbitNr = 74005447, BXID = 226, EventType = 3

L1ABC contains information about 6 previous and the current L1 accept, while TCDS L1history contains information for 16 previous L1 accepts. To get the information about current L1 in case of TCDS one needs to use TCDS methods:  getOrbitNr(),  getBXID(),  getEventType(). 
For the considered event, the values are: 
OrbitNr = 74005787, BXID = 3076, EventType = 1
in agreement with L1ABC AcceptOffset = 0 entry.

tsusa reviewed (commented):
The new EventWithHistory method with TCDS will not work correctly: the method should fill TinyEvent with the information about  current L1 and vector of TinyEvents with the information about the previous L1 accepted. Please correct it.

mmusich committed:
add a constructor that can take data members

mmusich committed:
do the loop on tcdsRecord.getFullL1aHistory, including the index=0

silviodonato commented:
Kind reminder @vischia

mmusich commented:
@vischia 
please consider these two additional commits: https://github.com/cms-sw/cmssw/commit/cf06a23c045a18f08df75ba72f747dbc2a767c04 and https://github.com/cms-sw/cmssw/commit/d508cc00d18654f3afbde55730dbc3fc2d314ab7. 
These should fix the problem reported by @tsusa  at https://github.com/cms-sw/cmssw/pull/29116#pullrequestreview-372586813. 
We've tested both the online and offline DQM clients without remaining warning messages.

vischia commented:
> @vischia which is your release base? That workflow is defined only in CMSSW >= 11_0_X (introduced for the first time in #27341)

@mmusich I am working in CMSSW_11_0_0_pre1, so maybe the workflow was not defined yet in the prerelease?

vischia commented:
> @vischia
> Below is the dump of L1ABC and TCDS L1history for Run 331483, Event 3668020
> L1ABC
> AcceptOffset = 0, OrbitNr = 74005787, BXID = 3076, EventType = 1
> AcceptOffset = -1, OrbitNr = 74005772, BXID = 2798, EventType = 1
> AcceptOffset = -2, OrbitNr = 74005770, BXID = 2786, EventType = 1
> AcceptOffset = -3, OrbitNr = 74005748, BXID = 743, EventType = 1
> AcceptOffset = -4, OrbitNr = 74005742, BXID = 2380, EventType = 1
> AcceptOffset = -5, OrbitNr = 74005740, BXID = 3490, EventType = 1
> AcceptOffset = -6, OrbitNr = 74005719, BXID = 2537, EventType = 1
> AcceptOffset = -7, OrbitNr = 74005699, BXID = 1941, EventType = 1
> TCDS L1history
> index = -1, OrbitNr = 74005772, BXID = 2798, EventType = 3
> index = -2, OrbitNr = 74005770, BXID = 2786, EventType = 3
> index = -3, OrbitNr = 74005748, BXID = 743, EventType = 3
> index = -4, OrbitNr = 74005742, BXID = 2380, EventType = 3
> index = -5, OrbitNr = 74005740, BXID = 3490, EventType = 2
> index = -6, OrbitNr = 74005719, BXID = 2537, EventType = 3
> index = -7, OrbitNr = 74005699, BXID = 1941, EventType = 3
> index = -8, OrbitNr = 74005694, BXID = 3039, EventType = 3
> index = -9, OrbitNr = 74005686, BXID = 3506, EventType = 3
> index = -10, OrbitNr = 74005656, BXID = 3325, EventType = 3
> index = -11, OrbitNr = 74005628, BXID = 3490, EventType = 2
> index = -12, OrbitNr = 74005558, BXID = 3021, EventType = 3
> index = -13, OrbitNr = 74005517, BXID = 3263, EventType = 3
> index = -14, OrbitNr = 74005516, BXID = 3490, EventType = 2
> index = -15, OrbitNr = 74005459, BXID = 3293, EventType = 3
> index = -16, OrbitNr = 74005447, BXID = 226, EventType = 3
> 
> L1ABC contains information about 7 previous and the current L1 accept, while TCDS L1history contains information for 16 previous L1 accepts. To get the information about current L1 in case of TCDS one needs to use TCDS methods: getOrbitNr(), getBXID(), getEventType().
> For the considered event, the values are:
> OrbitNr = 74005787, BXID = 3076, EventType = 1
> in agreement with L1ABC AcceptOffset = 0 entry.

Hi @tsusa , thank you very much for the test; indeed it also clarified to me a couple details.

mmusich commented:
> @mmusich I am working in CMSSW_11_0_0_pre1, so maybe the workflow was not defined yet in the prerelease?

yes. It entered in pre-3. https://github.com/cms-sw/cmssw/releases/tag/CMSSW_11_1_0_pre3.
I strongly advise you to rebase to master....

vischia commented:
> Kind reminder @vischia

Hi @silviodonato , I had warned about one week or so of expected inactivity on this.

vischia commented:
> @vischia
> please consider these two additional commits: [cf06a23](https://github.com/cms-sw/cmssw/commit/cf06a23c045a18f08df75ba72f747dbc2a767c04) and [611af70](https://github.com/cms-sw/cmssw/commit/611af70cd8c96cd5c96350aad3f1f82aae407bb6).
> These should fix the problem reported by @tsusa at [#29116 (review)](https://github.com/cms-sw/cmssw/pull/29116#pullrequestreview-372586813).
> We've tested both the online and offline DQM clients without remaining warning messages.

Hi @mmusich, thank you! I'll pull them, look into the changes to understand them, and update the PR.

vischia committed:
Merge remote-tracking branch 'mmusich/fixTCDs_forPietro' into tdcsScalTicket

cmsbuild commented:
The code-checks are being triggered in jenkins.

vischia commented:
Rebased to cmssw:master, and merged @mmusich 's commits

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14292

- This PR adds an extra 116KB to repository

cmsbuild commented:
Pull request #29116 was updated. @perrotta, @emeschi, @cmsbuild, @slava77, @mommsen, @santocch can you please check and sign again.

mommsen commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **MATRIX_EXTRAS** = 138.1

cmsbuild commented:
-1

Tested at: 9cb7235ee1351aed1f9a8f2a15c9b6d435e495f0

CMSSW: CMSSW_11_1_X_2020-03-20-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-524d9c/5292/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
11634.0 step3
<pre>runTheMatrix-results/11634.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2021_GenSimFull+DigiFull_2021+RecoFull_2021+HARVESTFull_2021+ALCAFull_2021/step3_TTbar_14TeV+TTbar_14TeV_TuneCP5_2021_GenSimFull+DigiFull_2021+RecoFull_2021+HARVESTFull_2021+ALCAFull_2021.log

</pre>12434.0 step3
<pre>runTheMatrix-results/12434.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2023_GenSimFull+DigiFull_2023+RecoFull_2023+HARVESTFull_2023+ALCAFull_2023/step3_TTbar_14TeV+TTbar_14TeV_TuneCP5_2023_GenSimFull+DigiFull_2023+RecoFull_2023+HARVESTFull_2023+ALCAFull_2023.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

mmusich commented on DPGAnalysis/SiStripTools/plugins/APVCyclePhaseProducerFromL1TS.cc:
```diff
@@ -185,38 +190,43 @@ void APVCyclePhaseProducerFromL1TS::produce(edm::Event& iEvent, const edm::Event
-      orbitoffset = _useEC0 ? (*l1ts)[0].lastEventCounter0() + _magicOffset : (*l1ts)[0].lastResync() + _magicOffset;
+  if (useTCDS) {
+    // l1ts->empty() always retuns false (last commit as of today: https://github.com/cms-sw/cmssw/commit/f4694d795d4b268d541c633dfb68283d889264b0 ), so the check is likely not necessary---and TCDSRecord hasn't anything similar
+    if (tcdsRecord->getLastResync() != 0) {
```
@vischia @mommsen 
I investigated a bit the error reported by the bot [here](https://github.com/cms-sw/cmssw/pull/29116#issuecomment-601833108) and I found out that this is the line triggering the problem reported:
```
----- Begin Fatal Exception 23-Mar-2020 11:45:26 CET-----------------------
An exception of category 'StdException' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Prefetching for module SiStripMonitorCluster/'HLTSiStripMonitorCluster'
   [2] Calling method for module APVCyclePhaseProducerFromL1TS/'APVPhases'
Exception Message:
A std::exception was thrown.
vector::_M_range_check: __n (which is 5) >= this->size() (which is 0)
----- End Fatal Exception -------------------------------------------------
```

As far as I can see `getLastResync()` is implemented here:

https://github.com/cms-sw/cmssw/blob/1711d0fe46c07e72b3a77242f61df6fe98860fad/DataFormats/TCDS/interface/TCDSRecord.h#L121

and indeed `BGo::Resync` is the fifth element of this enum:

https://github.com/cms-sw/cmssw/blob/1711d0fe46c07e72b3a77242f61df6fe98860fad/DataFormats/TCDS/interface/TCDSRecord.h#L27

It seems to me that the `lastBgos_` vector is empty.
The two failing workflows (`11634.0` and `12434.0`) are both Run3 simulation workflows, and therefore in both cases  `forceSCAL` is set to `False`. 
What should we expect for the TCDS unpacker in the simulation case? Is this vector supposed to be filled?

venturia replied:
If the issue is with simulation would suggest to adopt the same strategy adopted for SCAL. Also in that case no one expected that the simulation provided sensible values.

mmusich replied:
@venturia I tested this 
```suggestion
    if (tcdsRecord->getLastResync() != 0 && iEvent.eventAuxiliary().isRealData()) {
```
and I confirm it works (at least it does not crash). Though I would like the DAQ team to confirm that's the expected behaviour.

mommsen replied:
There's indeed nothing sensible to be simulated for the TCDS payload. Thus, you should rely upon TCDS information only for real data.

mmusich replied:
Thank you @mommsen 
@vischia can you implement what has been suggested above?

vischia replied:
@mmusich @mommsen @davidlange6 Done, but in the version suggested by @mommsen 

```
 if (useTCDS && iEvent.eventAuxiliary().isRealData()) {
```
because it makes sense to me that the check is performed for all the conditionals inside that if.

@davidlange6 was suggesting to add the condition only to the inner if in https://github.com/cms-sw/cmssw/pull/29116#issuecomment-602562860

```
if (tcdsRecord->getLastResync() != 0 && iEvent.eventAuxiliary().isRealData()) {
```
and reports it does not crash, so maybe on the other side it is not necessary to have the condition in the outer loop. Please let me know if that's the case (I implemented the one suggested by @mommsen , am pushing now)

mmusich replied:
@vischia, David was just copying the earlier version of my own suggestion, which was buggy. 
Please go ahead with the solution I suggested here: https://github.com/cms-sw/cmssw/pull/29116#discussion_r396394638

vischia replied:
@mmusich OK perfect, that's what I uploaded indeed. Thanks for confirming!

davidlange6 commented:
Checking for real-data first would be more clear (and perhaps more robust)



@venturia<https://github.com/venturia> I tested this

⬇️ Suggested change

-    if (tcdsRecord->getLastResync() != 0) {
+    if (tcdsRecord->getLastResync() != 0 && iEvent.eventAuxiliary().isRealData()) {


and I confirm it works (at least it does not crash). Though I would like the DAQ team to confirm that's the expected behaviour.

vischia committed:
Fix for when TCDS is run on MC (where no meaningful info is present)

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29116/14328

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #29116 was updated. @perrotta, @emeschi, @cmsbuild, @slava77, @mommsen, @santocch can you please check and sign again.

pieterdavid commented:
I have not followed very closely, but with the last commit by @vischia the outstanding issues should be resolved, right? The workflows 11634.0 and 12434.0 run without errors for me with the latest version, could someone start the tests? Thanks in advance

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **MATRIX_EXTRAS** = 138.1

cmsbuild commented:
+1
Tested at: bae14a9a57951d05e0330e21c666ab780fdd1ffe
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-524d9c/5356/summary.html
CMSSW: CMSSW_11_1_X_2020-03-24-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-524d9c/5356/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-524d9c/138.1_RunCosmics2020+RunCosmics2020+RECOCOSDRUN3+ALCACOSDRUN3+HARVESTDCRUN3

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2692493
 - DQMHistoTests: Total failures: 35
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2692139
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

vischia commented:
Hi,

I see the tests passed, I suppose this is then ready to be merged, right?

Cheers,
Pietro

mmusich commented:
@vischia several people still need to sign...

mommsen commented:
+1

perrotta commented:
+1
- Reco only involved with DataFormats/TCDS/interface/L1aInfo.h, where a ctor with explicit parameter settings is added
- Jenkins tests pass
- Differences in DQM (L1T) for wf 25202 correspond to what reported in the pending issue #29237, and are therefore not related to this PR

25202.0
--




25202.0
--

santocch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @silviodonato, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
